src/main/scala/io/iohk/ethereum/blockchain/sync/fast/FastSyncBranchResolverActor.scala

```diff
@@ -0,0 +1,273 @@
+package io.iohk.ethereum.blockchain.sync.fast
+
+import akka.actor.{Actor, ActorLogging, ActorRef, PoisonPill, Props, Scheduler, Terminated}
+import io.iohk.ethereum.blockchain.sync.Blacklist.BlacklistReason
+import io.iohk.ethereum.blockchain.sync.PeerListSupportNg.PeerWithInfo
+import io.iohk.ethereum.blockchain.sync.PeerRequestHandler.{RequestFailed, ResponseReceived}
+import io.iohk.ethereum.blockchain.sync.fast.FastSyncBranchResolverActor._
+import io.iohk.ethereum.blockchain.sync.{Blacklist, PeerListSupportNg, PeerRequestHandler}
+import io.iohk.ethereum.db.storage.AppStateStorage
+import io.iohk.ethereum.domain.{BlockHeader, Blockchain}
+import io.iohk.ethereum.network.Peer
+import io.iohk.ethereum.network.p2p.messages.Codes
+import io.iohk.ethereum.network.p2p.messages.PV62.{BlockHeaders, GetBlockHeaders}
+import io.iohk.ethereum.utils.Config.SyncConfig
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.duration._
+import monix.eval.Coeval
+import monix.catnap.cancelables.AssignableCancelableF.Bool
+import akka.actor.Timers
+import io.iohk.ethereum.blockchain.sync.fast.BinarySearchSupport.ContinueBinarySearch
+
+class FastSyncBranchResolverActor(
+    val fastSync: ActorRef,
+    val peerEventBus: ActorRef,
+    val etcPeerManager: ActorRef,
+    val blockchain: Blockchain,
+    val blacklist: Blacklist,
+    val syncConfig: SyncConfig,
+    val appStateStorage: AppStateStorage,
+    implicit val scheduler: Scheduler
+) extends Actor
+    with ActorLogging
+    with FastSyncBranchResolver
+    with PeerListSupportNg
+    with Timers {
+
+  import FastSyncBranchResolverActor._
+  import FastSyncBranchResolver._
+  import BinarySearchSupport._
+
+  private val recentHeadersSize: Int = syncConfig.blockHeadersPerRequest
+
+  private val recentBlocksSearch: RecentBlocksSearch = new RecentBlocksSearch(blockchain)
+
+  override def receive: Receive = waitingForPeerWithHighestBlock
+
+  private def waitingForPeerWithHighestBlock: Receive = handlePeerListMessages orElse { case StartBranchResolver =>
+    getPeerWithHighestBlock match {
+      case Some(PeerWithInfo(peer, _)) => requestRecentBlockHeaders(peer, blockchain.getBestBlockNumber())
+      case None =>
+        log.info(s"Waiting for peers, rescheduling StartBranchResolver")
```

Comment:
Could you please clean the string interpolations that are not necessary? I have seem so many in the code base in general

Reply:
removed this one...there's already a ticket to add a linter to the codebase to catch unused things